VPN-5860: Improve health metrics

mcleinman opened the pull request:
## Description

The first round of health metrics from Super Dooper dashboards needed improvement:
- We were spamming Glean with events when connections were down, which was too noisy for useful engineering and data insights.
- It was still hard to answer questions like "how long are connections stable for?"

This PR attempts to address this. The current solution was designed in collaboration w/ Yeonjoo (data) and Santiago (product). We're adding 9 new metrics, and removing a bunch as well. We think that some subset of these 9 will be most useful, but aren't quite sure. Thus, these metrics all will live for about 6 months, and we'll then evaluate the data and decide which ones to keep in perpetuity.

These metrics are reworked both in the app and in the Android daemon.

The metrics (3 of each, for stable, unstable, and no connection):
- An event for when the connectivity *changes* to this state.
- A counter for when a health check returns this state, even if it was already in the state.
- A timing distribution for how long it stays in a given state. (This only exists on Android daemon and desktop, as the mobile apps are frequently relaunched during a VPN session.)

The counters can be compared in relative terms, but should not be considered as markers of time. There is a race condition in the counter, and it cannot be relied upon as having one count per second. In many cases, we conduct a health check due to the timer timeout. Then milliseconds later, a ping returns and we do a health check due to that ping returning. However, if the ping returned before the timeout, the timer would have been cancelled we would have only recorded one. Fixing this is outside the scope of this PR, and the resulting data is still valuable.

I also renamed a couple Android booleans, to make it clear when the silent server switch (or "reconnect") is coming from the app, and when it's coming from the daemon.

I've tested this on macOS, Android, and iOS.

## Reference

https://mozilla-hub.atlassian.net/browse/VPN-5860?focusedCommentId=825842

## Checklist
    
- [x] My code follows the style guidelines for this project
- [x] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [x] I have performed a self review of my own code
- [x] I have commented my code PARTICULARLY in hard to understand areas
- [x] I have added thorough tests where needed

mcleinman commented:
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

(2 of these metrics were [previously approved](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/5702#issuecomment-1397761161), so this form is for the one new one.)

1) What questions will you answer with this data?

These are a handful of metrics to learn more about how often VPN sessions have connectivity issues. How many times in a VPN session do users have connectivity issues? How long do those issues last for?

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? 

We do not have great insight into some of basic questions about our users: How often are they having issues? What sort of issues are they? How long do users keep the VPN active for?

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

Carefully considered which metrics were needed. Additionally had some prior metrics (which are removed in this PR), but they were not robust enough to answer the questions above.

4) Can current instrumentation answer these questions?

No, unfortunately. See question 3 for more details.

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   

**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**

<table>
  <tr>
    <td>Measurement Name</td>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>changed_to_no_signal</td>
    <td>Event that is recorded when connection status changes to no signal</td>
    <td>technical</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-5860">VPN-5860</a></td>
  </tr>
  <tr>
    <td>changed_to_unstable</td>
    <td>Event that is recorded when connection status changes to unstable</td>
    <td>technical</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-5860">VPN-5860</a></td>
  </tr>
  <tr>
    <td>changed_to_stable</td>
    <td>Event that is recorded when connection status changes to stable</td>
    <td>technical</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-5860">VPN-5860</a></td>
  </tr>
  <tr>
    <td>no_signal_count</td>
    <td>Count of health checks which result in status no signal</td>
    <td>technical</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-5860">VPN-5860</a></td>
  </tr>
  <tr>
    <td>unstable_count</td>
    <td>Count of health checks which result in status unstable</td>
    <td>technical</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-5860">VPN-5860</a></td>
  </tr>
  <tr>
    <td>stable_count</td>
    <td>Count of health checks which result in status stable</td>
    <td>technical</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-5860">VPN-5860</a></td>
  </tr>

  <tr>
    <td>no_signal_time</td>
    <td>Time distribution of how long connectivity status stays in no signal</td>
    <td>technical</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-5860">VPN-5860</a></td>
  </tr>
  <tr>
    <td>unstable_time</td>
    <td>Time distribution of how long connectivity status stays in unstable</td>
    <td>technical</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-5860">VPN-5860</a></td>
  </tr>
  <tr>
    <td>stable_time</td>
    <td>Time distribution of how long connectivity status stays in stable</td>
    <td>technical</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-5860">VPN-5860</a></td>
  </tr>
</table>

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.

This collection is documented in the Glean Dictionary at https://dictionary.telemetry.mozilla.org/

7) How long will this data be collected?  Choose one of the following:

I want this data to be collected for 6 months initially (potentially renewable). 
(It's likely that after reviewing 6 months of data, a subset of these metrics will be collected in perpetuity. And some will be removed)

8) What populations will you measure?

No filters - all channels, countries, and locales, unless the user has opted out of data collection on that device.

9) If this data collection is default on, what is the opt-out mechanism for users?

When launching the app for the first time, a user is given an option for whether to allow data collection. After this initial set up screen, a user can always toggle data collection permissions in the System Preferences screen.

10) Please provide a general description of how you will analyze this data.

Dashboards that product managers and others will consult on a regular basis.

11) Where do you intend to share the results of your analysis?

Within the Mozilla VPN team.

12) Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection? 

No

mcleinman commented:
Tagging @travis79 for data-review

mcleinman commented:
> Please add unit tests for all of the new collections or add to the description of this ticket why that is not possible at the moment with a follow up ticket for implementation of tests in the future.

For Android daemon, quoting [a prior PR](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/9046#issuecomment-1916943291): "We cannot test classes that load ndk-based native code. So neither vpnservice (wireguard.so) and glean are good candidates for this type of tests." There is [an existing ticket](https://mozilla-hub.atlassian.net/browse/VPN-2391) for this work.

For the app (and the daemon, as well), we don't have any unit tests that activate/deactivate the VPN, which is needed to test these. There is [an existing ticket](https://mozilla-hub.atlassian.net/browse/VPN-4453) for better testing of this part of the codebase.

brizental commented:
For the Android part, sorry I forgot I asked the exact same thing on your previous PR 😅 

However for the C++ code, given we are not adding the new collections to the controller.cpp file itself (which is mocked in tests), I think it is worth attempting to unit test the new stuff still -- I believe it is possible. 

> For the app (and the daemon, as well), we don't have any unit tests that activate/deactivate the VPN, which is needed to test these.

Given the new code here is not in controller.cpp specifically, but on connectionhealth.cpp and telemetry.cpp, I think it is still worth attempting to unit test it. The ConnectionHealth is a testable class with examples: https://github.com/mozilla-mobile/mozilla-vpn-client/blob/main/tests/unit/testconnectionhealth.cpp. I believe telemetry.cpp would be testable as well, although there are no tests for it at the moment. I'd be happy to pair on this if you'd like.

brizental commented on src/telemetry/metrics.yaml:
```diff
@@ -1776,3 +1697,241 @@ session:
+      Count of times that the connection health check results
+      in no signal.
+
+      The health check counters can be compared in relative terms, but should not be considered as
+      markers of time. There is a race condition in the health check (which calls the telemetry),
+      and we frequently have more than one count per second. This is because we conduct a health check
+      due to the timer timeout, and milliseconds later a ping returns and we do another health check
+      due to that ping returning. 
```
I'd suggest using a different terminology than `pings` here or maybe adding a note about what is meant by this term in this paragraph. In telemetry documentation that will usually be read as a Glean ping (at least that is what I read).

mcleinman replied:
Good call. I rewrote this, let me know if you think it still needs more work.

brizental commented on src/telemetry/metrics.yaml:
```diff
@@ -1776,3 +1697,241 @@ session:
+      in unstable.
+
+      The health check counters can be compared in relative terms, but should not be considered as
+      markers of time. There is a race condition in the health check (which calls the telemetry),
```
Thought: is this a race condition or just expected behaviour?

mcleinman replied:
A bit of both? Because if that network check ping returned before the timer timed out, we'd only have 1 health check. And if the ping never returned, we'd only have 1 health check. This could be considered an expected edge case, so I'm removing the "race condition" language.

brizental commented on src/telemetry/metrics.yaml:
```diff
@@ -1776,3 +1697,241 @@ session:
+    description: |
+      Count of times that the connection health check succeeds.
+
+      The health check counters can be compared in relative terms, but should not be considered as
```
Everytime I re-read this sentence I have new questions, sorry :sweat_smile: What do you mean here by "relative terms"? FWIW, I think the callout about not considering counters as markers of time is unnecessary. Counters are never markers of time, anyways.

mcleinman replied:
Agree they're not markers of time, but I could see down the line someone asking "how often we record the counter" and a quick glance of the code saying "every 1 second". I think it's important to underscore here.

By "relative terms" I meant "you can compare the ratio of the 3 counters" but I'm removing that because I no longer think that is true - based on what I wrote above (about sometimes getting 1 health check and sometimes getting 2 - depending on when the network ping returned).

brizental commented on src/telemetry/metrics.yaml:
```diff
@@ -1776,3 +1697,241 @@ session:
+      - vpnsession
+      - daemonsession
+    description: |
+      Lengths of times spent in no signal state.
```
nit, applies to others:

```suggestion
      Time spent  in no signal state.
```

mcleinman replied:
Updated, thanks.

brizental commented on src/telemetry/metrics.yaml:
```diff
@@ -1776,3 +1697,241 @@ session:
+
+      Only collected on desktop for vpnsession, as mobile apps
+      frequently are relaunched during VPN sessions. It is
+      collected in daemonsessino, currently only on Android.
```
```suggestion
      collected in daemonsession, currently only on Android.
```

brizental replied:
This errors happens in other places as well.

mcleinman replied:
Egg on my face. Thanks, updated.

brizental commented on tests/unit/testconnectionhealth.cpp:
```diff
@@ -62,20 +75,89 @@ void TestConnectionHealth::healthCheckup() {
+
+  // test the 3 timespans
+
+  // .count is always returning 0, so commenting this out for now.
```
Hm, thanks for filling a ticket! This is looking like a Glean bug more than a bug on our side. For the time being, something like

```
int getTimingDistCountFromValues(QHash<int, int> values) {
    int count = 0;
    for (auto i = values.constBegin(); i != values.constEnd(); ++i) {
      count += i.value();
    }

    return count;
}
```

And then called like: 

```
getTimingDistCountFromValues(mozilla::glean::connection_health::stable_time.testGetValue().values)
```

Should work.

brizental replied:
Nah, bug on our side: https://github.com/mozilla/glean/blob/main/glean-core/src/metrics/mod.rs#L89-L91, count is skipped when serializing this structure to JSON on the rust side :facepalm: Anyways, that workaround should do for now. I'll fix the ticket at some point.

mcleinman replied:
Updated, thanks.

mcleinman commented on src/telemetry/metrics.yaml:
```diff
@@ -1776,3 +1697,241 @@ session:
+      done on the Android daemon and not on the iOS network extension.
+
+      A no signal event event happens when the connection is showing issues,
+      and a silent switch will not address the issue. On when the connection
```
🤦🏻 👍🏻

mcleinman commented on src/telemetry/metrics.yaml:
```diff
@@ -1776,3 +1697,241 @@ session:
+      is not the same. On the daemon this checks are only
+      done on the Android daemon and not on the iOS network extension.
+
+      A no signal event event happens when the connection is showing issues,
```
🤦🏻👍🏻

mcleinman commented on tests/unit/testconnectionhealth.h:
```diff
@@ -2,17 +2,24 @@
   void dnsPingReceived();
   void healthCheckup();
   void updateDnsPingLatency();
+  void metricsTest(int expectedStablePeriods, int expectedUnstablePeriods,
```
Yup, thanks.

mcleinman commented on tests/unit/testconnectionhealth.cpp:
```diff
@@ -62,20 +75,89 @@ void TestConnectionHealth::healthCheckup() {
            ConnectionHealth::ConnectionStability::Stable);
+  metricsTest(2, 2, 1, ConnectionHealth::ConnectionStability::Stable);
+}
+
```
Good call. Moved it to it's own test.

mcleinman commented on src/connectionhealth.cpp:
```diff
@@ -275,3 +282,84 @@ void ConnectionHealth::overwriteStabilityForInspector(
   emit stabilityChanged();
 }
+
+void ConnectionHealth::startMetricsTimer(ConnectionStability stability) {
```
Renamed these and the `stopMetricsTimer`s

mcleinman commented on src/connectionhealth.cpp:
```diff
@@ -275,3 +282,84 @@ void ConnectionHealth::overwriteStabilityForInspector(
+  logger.debug() << "MATTHEW starting metrics outside";
+#if defined(MZ_WINDOWS) || defined(MZ_LINUX) || defined(MZ_MACOS) || \
+    defined(MZ_DUMMY)
+  logger.debug() << "MATTHEW starting metrics inside";
```
🤦🏻 That's embarrassing. Done.

brizental replied:
We've all been there :fist_raised:

mcleinman commented on src/connectionhealth.cpp:
```diff
@@ -119,18 +123,20 @@ void ConnectionHealth::setStability(ConnectionStability stability) {
   }
 
+  // Pings will sometimes come between VPN sessions, triggering setStability. We
+  // do not want to record count metrics in these cases.
```
Added clarifying language. 😅

mcleinman commented on android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNService.kt:
```diff
@@ -60,15 +60,15 @@ class VPNService : android.net.VpnService() {
             return (this.mConfig?.optInt("reason") ?: 0) == 1
         }
 
     private val shouldRecordMetrics: Boolean
```
That's a great call out, thanks. I just renamed these and added some more comments. Let me know if it's still confusing, and I'll improve it further.

mcleinman commented on android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNService.kt:
```diff
@@ -60,15 +60,15 @@ class VPNService : android.net.VpnService() {
         }
 
-    private val isChangingServers: Boolean
+    private val isAppChangingServers: Boolean
```
Good call, added comment.

mcleinman commented on src/telemetry/metrics.yaml:
```diff
@@ -1776,3 +1697,241 @@ session:
+      done on the Android daemon and not on the iOS network extension.
+
+      Stable status is when pings are returning in an expected
+      timeframe.
```
Great call, updated.

mcleinman commented:
@brizental  I believe I've addressed all your comments.

Testing the number of timespans fails about 10% of the time. When it fails, it's because the test setup has created an extra stable period. I'm not quite sure why this is happening. You'll see I've removed a silent server switch when testing, but I can't figure out why else this might be happening. From using the Glean debug viewer, I'm yet to see any instances of this happening in the actual app - My hunch is that this is a testing setup peculiarity. Do you see anything that might be causing this?

mcleinman commented on android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNService.kt:
```diff
@@ -313,7 +315,15 @@ class VPNService : android.net.VpnService() {
         mConfig = json
+        // shouldRecordStartTelemetry must be calculated after mConfig is set (on prior line)
+        // We don't want to record start metrics in several situations:
+        // - If Super Dooper feature is not active. (Covered by shouldRecordTimerAndEndMetrics)
```
Probably not at this point, unless we want the ability to turn it off quickly.

mcleinman commented on src/connectionhealth.cpp:
```diff
@@ -119,19 +123,24 @@ void ConnectionHealth::setStability(ConnectionStability stability) {
     return;
   }
 
+  // Connection check pings sometimes come between VPN sessions, triggering
```
I can't say for sure, but it was something I noticed in testing.

My guess is these are pings sent, then session ends, then pings return. But I didn't investigate.

brizental replied:
Ok, that makes sense.

mcleinman commented:
Working on redoing these tests, but need to close the laptop. Will finish this up tomorrow.

mcleinman commented on src/connectionhealth.cpp:
```diff
@@ -119,19 +123,24 @@ void ConnectionHealth::setStability(ConnectionStability stability) {
+    recordMetrics(m_stability, stability);
+  }
+
+// Silent switch adds non-deterministic metrics when testing.
```
I removed this. This was an artifact of the old way we were running tests on this, and no longer exists with the current approach.

mcleinman commented on src/telemetry/metrics.yaml:
```diff
@@ -1776,3 +1697,241 @@ session:
+
+      An unstable event happens when the connection is showing issues,
+      but a silent switch may still address the issue and no previous
+      silent switches have happened yet.
```
Updated, thanks.

mcleinman commented on tests/unit/testconnectionhealth.cpp:
```diff
@@ -78,4 +87,121 @@ void TestConnectionHealth::healthCheckup() {
 
-static TestConnectionHealth s_testConnectionHealth;
\ No newline at end of file
+void TestConnectionHealth::testTelemetry() {
```
Thanks, did a bunch of work on the tests. That first draft was based on the `healthCheckup` tests. I think it's now a bit clearer.

mcleinman commented on src/connectionhealth.cpp:
```diff
@@ -119,19 +123,24 @@ void ConnectionHealth::setStability(ConnectionStability stability) {
+#if defined(UNIT_TEST)
+      // Always record metrics when testing
+      state == Controller::StateInitializing ||
+#endif
```
Thanks, did a bunch of work on the tests. That first draft was based on the `healthCheckup` tests. I think it's now a bit clearer. And it now includes a test when the controller is not activated.

mcleinman commented:
@brizental I believe I've addressed all comments, and this is ready for another round of review.

brizental commented on tests/unit/testconnectionhealth.cpp:
```diff
@@ -78,4 +88,118 @@ void TestConnectionHealth::healthCheckup() {
+void TestConnectionHealth::testTelemetry() {
+  ConnectionHealth connectionHealth;
+
+  // Nothing recorded at start.
+  metricsTestErrorAndChange(0, 0, 0);
+  metricsTestCount(0, 0, 0);
+  metricsTestTimespan(0, 0, 0);
+
+  // Shouldn't do anything if controller state isn't on.
+  connectionHealth.startActive("", "");
+  connectionHealth.stop();
+  metricsTestErrorAndChange(0, 0, 0);
+  metricsTestCount(0, 0, 0);
+  metricsTestTimespan(0, 0, 0);
+
+  TestHelper::controllerState = Controller::StateOn;
+
+  connectionHealth.setStability(ConnectionHealth::Unstable);
+  connectionHealth.startActive("", "");
+  metricsTestErrorAndChange(0, 1, 0);
+  metricsTestCount(0, 1, 0);
+  metricsTestTimespan(0, 0, 0);
+
+  connectionHealth.setStability(ConnectionHealth::ConnectionStability::Stable);
+  metricsTestErrorAndChange(1, 1, 0);
+  metricsTestCount(1, 1, 0);
+  metricsTestTimespan(0, 1, 0);
+
+  connectionHealth.setStability(
+      ConnectionHealth::ConnectionStability::NoSignal);
+  metricsTestErrorAndChange(1, 1, 1);
+  metricsTestCount(1, 1, 1);
+  metricsTestTimespan(1, 1, 0);
+
+  // stops and sets to stable
+  connectionHealth.stop();
+  metricsTestErrorAndChange(2, 1, 1);
+  metricsTestCount(2, 1, 1);
+  metricsTestTimespan(1, 1, 1);
```
Would you mind adding a comment for each of these states? Like you do on line 99? The test becomes a but confusing because of the anonymous arguments for the tester fucntions.

mcleinman replied:
Comments added, thanks.

brizental replied:
Maybe I am missing something but I don't see them?

mcleinman replied:
🤦🏻 I forgot to push up the commit. You should see it now.

brizental commented on src/connectionhealth.cpp:
```diff
@@ -275,3 +279,84 @@ void ConnectionHealth::overwriteStabilityForInspector(
+void ConnectionHealth::stopTimingDistributionMetric(
+    ConnectionStability stability) {
+#if defined(MZ_WINDOWS) || defined(MZ_LINUX) || defined(MZ_MACOS) || \
+    defined(MZ_DUMMY)
```
nit: Maybe we can define a macro like "MZ_MOBILE" instead that is true when we are in IOS and ANDROID instead of this?

mcleinman replied:
Done, thanks.

brizental commented on src/telemetry/metrics.yaml:
```diff
@@ -1776,3 +1697,240 @@ session:
+    description: |
+      Time spent in no signal state.
+
+      Only collected on desktop for vpnsession, as mobile apps
+      frequently are relaunched during VPN sessions. It is
+      collected in daemonsession, currently only on Android.
```
If there are plans to add this to iOS it would be worth adding a ticket reference here.

brizental replied:
This applies to all comments where we say "currently only on Android" or similar.

mcleinman replied:
Added ticket in there
